### PR TITLE
Add IgnoredArguments to IVariableDeclarator.

### PR DESCRIPTION
**Pull request by 333fred**
Fixes part of https://github.com/dotnet/roslyn/issues/22175. The only tests still failing after this change are:
```
Microsoft.CodeAnalysis.CSharp.UnitTests.OutVarTests.GetEnclosingBinderInternalRecovery_09 FAILED
Microsoft.CodeAnalysis.CSharp.UnitTests.OutVarTests.GetEnclosingBinderInternalRecovery_10 FAILED
Microsoft.CodeAnalysis.CSharp.UnitTests.OutVarTests.GetEnclosingBinderInternalRecovery_11 FAILED
```
9 and 10 fail because we don't bind anything inside array rank specifiers in typeof operations, ie `typeof(int[10])`. 11 is because we don't do anything for this foreach scenario: `foreach other()`. I've opened #23006 and #23007 to track each of those problems.

vso: https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/520819

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs [hunk 1]
```diff
@@ -37,6 +37,7 @@ static void Main(string[] args)
         Conversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
         Operand: 
           ILocalReferenceOperation: o1 (OperationKind.LocalReference, Type: System.Object) (Syntax: 'o1')
+  IgnoredArguments(0)
```
Comment (heejaechang) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs]: can we not show this if IgnoredArguments is 0 ?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs]: We show IgnoredCondition on IDoLoop even when null, so I did the same here.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs]: This affects too many tests, which increases the cost of review. It would be good to keep the number of affected tests to a minimum.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 2]
```diff
@@ -5438,6 +5440,13 @@ public override IEnumerable<IOperation> Children
                 {
                     yield return Initializer;
                 }
+                if (!IgnoredArguments.IsEmpty)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >if (!IgnoredArguments.IsEmpty) [](start = 16, length = 30)

I would return them before the Initializer because syntactically they come first.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 3]
```diff
@@ -5472,14 +5483,17 @@ internal sealed partial class VariableDeclarator : BaseVariableDeclarator
...
     {
         private readonly Lazy<IVariableInitializerOperation> _lazyInitializer;
+        private readonly Lazy<ImmutableArray<IOperation>> _lazyIgnoredArguments;
 
-        public LazyVariableDeclarator(ILocalSymbol symbol, Lazy<IVariableInitializerOperation> initializer, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
+        public LazyVariableDeclarator(ILocalSymbol symbol, Lazy<IVariableInitializerOperation> initializer, Lazy<ImmutableArray<IOperation>> ignoredArguments, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
             base(symbol, semanticModel, syntax, type, constantValue, isImplicit)
         {
             _lazyInitializer = initializer ?? throw new System.ArgumentNullException(nameof(initializer));
+            _lazyIgnoredArguments = ignoredArguments;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >ignoredArguments [](start = 36, length = 16)

Should this be checked for null similar to the ```initializer```

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 4]
```diff
@@ -410,6 +410,7 @@ public override void VisitVariableDeclarator(IVariableDeclaratorOperation operat
             LogCommonPropertiesAndNewLine(operation);
 
             Visit(operation.Initializer, "Initializer");
+            VisitArray(operation.IgnoredArguments, "IgnoredArguments", logElementCount: true);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >VisitArray(operation.IgnoredArguments, "IgnoredArguments", logElementCount: true); [](start = 12, length = 82)

Please consider not printing anything if there are no ignored arguments

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >if (!IgnoredArguments.IsEmpty) [](start = 16, length = 30)

Please add a targeted test for the Children property.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Done.

---
In reply to: [149214465](https://github.com/dotnet/roslyn/pull/23008#discussion_r149214465) [](ancestors = 149214465)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 4 above)
Comment (333fred) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Done.

---
In reply to: [149217073](https://github.com/dotnet/roslyn/pull/23008#discussion_r149217073) [](ancestors = 149217073)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConversionExpression.cs]: These will be removed from the next push.

---
In reply to: [149209977](https://github.com/dotnet/roslyn/pull/23008#discussion_r149209977) [](ancestors = 149209977)

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Done.

---
In reply to: [149217374](https://github.com/dotnet/roslyn/pull/23008#discussion_r149217374) [](ancestors = 149217374)

**Comment by 333fred** [comment]
@AlekseyTs @heejaechang addressed first-round feedback.

**Comment by natidea** [comment]
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/520819

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs [hunk 5]
```diff
@@ -515,6 +517,99 @@ void M()
...
+                //         int /*<bind>*/x[10]/*</bind>*/;
+                Diagnostic(ErrorCode.WRN_UnreferencedVar, "x").WithArguments("x").WithLocation(6, 23)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<VariableDeclaratorSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void IVariableDeclaration_InvalidIgnoredArguments_VerifyChildren()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: >IVariableDeclaration_InvalidIgnoredArguments_VerifyChildren [](start = 20, length = 59)

We should also cover cases when either or both children are null.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3). Please add tests for Children property for scenarios when either or both children are null.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 6]
```diff
@@ -5424,16 +5424,25 @@ internal abstract partial class BaseVariableDeclarator : Operation, IVariableDec
...
         /// Optional initializer of the variable.
         /// </summary>
         public IVariableInitializerOperation Initializer => Operation.SetParentOperation(InitializerImpl, this);
+        public ImmutableArray<IOperation> IgnoredArguments => Operation.SetParentOperation(IgnoredArgumentsImpl, this);
 
         public override IEnumerable<IOperation> Children
         {
             get
             {
+                if (!IgnoredArguments.IsEmpty)
```
Comment (cston) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: No need to check `IsEmpty`.

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/IVariableDeclaratorOperation.cs [hunk 7]
```diff
@@ -35,5 +37,11 @@ public interface IVariableDeclaratorOperation : IOperation
         /// to check in all locations. It is only possible to have initializers in both locations in VB invalid code scenarios.
         /// </remarks>
         IVariableInitializerOperation Initializer { get; }
+
+        /// <summary>
+        /// Additional arguments supplied to the declarator in error cases, ignored by the compiler. This only used for C# cases, of
```
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/IVariableDeclaratorOperation.cs]: Typo at the end of the line?

**Review by mavasani — APPROVED** [review]

**Review by cston — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 6)

**Comment by AlekseyTs** [comment]
@333fred Is this PR blocked on anything? It looks like you had all sign-offs for several hours.

**Comment by 333fred** [comment]
@AlekseyTs integration tests to finish. I can override the integration tests, but I want to investigate the hang first.
